org.palladiosimulator.analyzer.slingshot.behavior.spd/src/org/palladiosimulator/analyzer/slingshot/behavior/spd/interpreter/utils/TargetGroupUtils.java

```diff
@@ -1,51 +1,221 @@
+				() -> getAllContextsToConsider(competingConsumersGroup));
+	}
+
+	/**
+	 * Checks whether the container is part of the target group. The way of how it
+	 * is checked, and what "being part of the target group" means, depends on the
+	 * concrete type of the target group.
+	 * 
+	 * @param container   The container to check.
+	 * @param targetGroup The target group to consider.
+	 * @return true if the container is part of the target group.
+	 * 
+	 * @see #isContainerInCompetingConsumersGroup(ResourceContainer,
+	 *      CompetingConsumersGroup)
+	 * @see #isContainerInElasticInfrastructure(ResourceContainer,
+	 *      ElasticInfrastructure)
+	 * @see #isContainerInServiceGroup(ResourceContainer, ServiceGroup)
+	 */
 	public static boolean isContainerInTargetGroup(final ResourceContainer container, 
 												   final TargetGroup targetGroup) {
 		if (targetGroup instanceof final ElasticInfrastructure elasticInfrastructure) {
 			return isContainerInElasticInfrastructure(container, elasticInfrastructure);
-		} else if (targetGroup instanceof final ServiceGroup serviceGroup) {
+		}
+		if (targetGroup instanceof final ServiceGroup serviceGroup) {
 			return isContainerInServiceGroup(container, serviceGroup);
 		}
+		if (targetGroup instanceof final CompetingConsumersGroup competingConsumersGroup) {
+			return isContainerInCompetingConsumersGroup(container, competingConsumersGroup);
+		}
 		
 		return false;
 	}
 	
+	/**
+	 * Checks whether the assembly context is part of the target group in the case
+	 * if the target group is of type {@link ServiceGroup} or
+	 * {@link CompetingConsumersGroup}. The assembly contexts to consider are the
+	 * unit assemblies, as well as the replicated assemblies.
+	 * 
+	 * In case of the {@link ElasticInfrastructure}, it checks whether the context
+	 * is referencing a resource container is part of the resource environment
+	 * referenced by the target group.
+	 * 
+	 * @param context     The assembly context to check.
+	 * @param targetGroup The target group to consider.
+	 * @return True if the assembly context is part of the target group (or
+	 *         referneced by a container).
+	 */
+	public static boolean isAssemblyInTargetGroup(final AssemblyContext context, final TargetGroup targetGroup) {
+		if (targetGroup instanceof final ServiceGroup serviceGroup) {
+			return getAllContextsToConsider(serviceGroup).anyMatch(ac -> ac.getId().equals(context.getId()));
+		}
+		if (targetGroup instanceof final ElasticInfrastructure elasticInfrastructure) {
+			return elasticInfrastructure.getPCM_ResourceEnvironment().getResourceContainer_ResourceEnvironment()
+					.stream().anyMatch(rc -> getContainerRelatedToContext(context)
+							.anyMatch(rcp -> rcp.getId().equals(rc.getId())));
+		}
+		if (targetGroup instanceof final CompetingConsumersGroup competingConsumers) {
+			return getAllContextsToConsider(competingConsumers).anyMatch(ac -> ac.getId().equals(context.getId()));
```

Comment:
this is broken for me:
`context` is `Assembly_RabbitBroker` which makes sense, because the message queue is located at the broker, i.e. the measurement was taken at the broker.  
`getAllContextsToConsider` contains `Assembly_DataProcessing` which is unit/elements in the semantic configuration. 
and the it does not match :( 

i guess you need to get te broker instead of the uni/elements here?

Reply:
I changed the implementation such that the broker is now also part of the stream, but left the elements there as well. We might need to discuss whether such a situation can occur, where the TargetGroup is a CompetingConsumerGroup and the measurement comes from one of the elements instead of the broker